tripal/tripal
# Better Debugging of ChadoStorage for Field Development

# Tripal 4 Core Dev Task

Issue #1594 


### Tripal Version: 4.x

## Description


As we have been developing more fields, it's becomming apparent that ChadoStorage is too much of a black box at this point. For example, in cases where you've defined the properties as expected based on documentation but they still don't work --you just don't get your data... There is no way for you to see the queries created by chadostorage, get an idea of how chadostorage is processing your properties, etc. This basically means there is no way to debug your properties without adding debugging statements into chado storage yourself.

This PR adds a checkbox to the field settings form to enable debugging information for a field. This allows developers who are working on fields to target a specific field instance (i.e. their field attached to a specific content type with a specific set of settings) and receive the following debugging information to give them insight into how chado storage is interpreting the properties they gave them. This will help with detecting bugs both in fields being developed and in chadostorage itself!

Debugging information made available:

- The status of chado storage at the beginning of the process including the following information for each attached field instance:
  - field name, type and definition
  - field settings
  - the number of properties
  - an array with the following information for each property type:
    - the class name with namespace
    - the term attached
    - the storage settings set in the field
- The property values at the beginning of the process
- An array describing the records to be created. This is the result of buildChadoRecords and is used directly by chadostorage not generated for debugging. This is aggregated across all fields.
- The primary key values of the so-called base records as understood by chado storage. This is aggregated across all fields.
- The actual queries being executed by chado storage. There will be the record key and delta in the label. Unfortunately the placeholder values are not included due to Drupal bugs documented in the code.
- The values after the process is complete. These will be the values made available to the widget and formatter classes.

![screenshot](https://user-images.githubusercontent.com/1566301/260199436-7efc55fb-bd1b-4a62-9dc1-f30980fb67bf.png)

## Notes

- This debugging information is posted for load, insert, update, validate (+ delete/find once implemented)... 
  - Inserting results in both a load and insert
  - update though is particularly long with multiple loads and the update. This is how Drupal works for some reason 🤷‍♀️ 
  - there can be a lot of information which could be overwhelming
- Since multiple fields make up a single chado record, if you enable debugging for a single field you actually see information for all fields on that content type. I felt this was needed to truely debug
- The content listing page might time out if there are too many entries with debugging turned on as that triggers a full load for every entity on the page.
- The field debugger uses dpm so devel is required to use it 🤷‍♂️ we can't use html in drupal_set_message and I felt formatting was necessary for this to be usable at all.

## Testing?



1. Create a docker image and container on this branch. [See these instructions for how to do that.](https://tripaldoc.readthedocs.io/en/latest/contributing/docker-for-testing.html#testing-an-unmerged-pull-request)
2. Choose a content type and field to debug (e.g. gene, organism field)
3. Go to Tripal > Page Structure > Gene > Manage Fields
4. Click on edit beside the field you want to debug
5. Select the "Enable debugging" checkbox and save the field settings
6. Create a gene page by going to Tripal > Content > Add Tripal Content > Gene and filling out the form
7. When you click save you would expect to see debugging information.
8. You can then refresh the page and confirm you see debugging information for the load.
9. Then edit the page and confirm you see debugging information for the load multiple times and the update.
